### PR TITLE
Scan assets from disk

### PR DESCRIPTION
**Pull request by fmatosqg**
Opt in parameter will scan for all assets under given folder.

To do: Scenario when app is built, then user adds one more asset without touching pubspec.yaml. It's currently not refreshing bundle, but comparing timestamp for both pubspec and list of files under scanned path will be enough.

Covers part of #4890

### CODE REVIEW

**Comment by fmatosqg** [comment]
@tvolkert I'd appreciate help with the windows failed test, I'm stuck at it without access to windows machine. 

Can't reproduce on my OS X. Stack trace is not helpful. Should I add loads of print() to debug AppVeyor?

**Comment by tvolkert** [comment]
@fmatosqg actually, see the discussion in https://github.com/flutter/flutter/pull/14246.  The fix to google/file.dart#68 just landed (and as of https://github.com/flutter/flutter/pull/16442 is rolled into Flutter), so the right answer is actually to update the asset tests to use `MemoryFileSystem`, with a style of `FileSystemStyle.windows` for the windows tests.

/cc @goderbauer

**Comment by fmatosqg** [comment]
@tvolkert I'm going on a rabbit hole, I'd appreciate some guidance.

- Should I replace all instances I find of fs.file(<path>), fs.isFileSync(<path>) etc with something coming from MemoryFileSystem ? If yes, removing `import 'base/file_system.dart';` from all related files is the way to get started and find all offending uses?
- Should I use `FileSystemStyle.windows` for all build systems, and if not how would I do the check?

**Comment by tvolkert** [comment]
Sorry - I started reviewing the PR, and before we dive too deeply in fixing the tests, let's take a step back on the design.

Rather than having a `enable-asset-scan` attribute in the pubspec, I think we should just respect asset entries that follow a certain pattern, such as:

1. Any entry that ends with a slash is treated as a directory, and we include all files within the directory (recursively)
1. We support [globstar-style](https://www.gnu.org/software/bash/manual/html_node/The-Shopt-Builtin.html) syntax (the pattern ‘**’ used in a filename expansion context will match all files and zero or more directories and subdirectories.)

**Comment by fmatosqg** [comment]
@tvolkert Thanks for review, some feedback on feedback:

That makes sense, though I'm not sure how to interpret some entries with globstar on variants.

I'm not familiar with globstar so I'll draft some unit tests right here.

Consider these files:
```
a/file
a/b/file
a/b/c1/file
a/b/c2/file
```
If I aim at `a/` then we should get asset `a/file` with variant `a/b/file`
If I aim at `a/b/**/` it's still easy but tricky to implement, asset is `a/b/file` and it has 2 variants `a/b/c1/file` and `a/b/c2/file`
Now if we do `a/**/` things get messy and I don't know how to tell variants from main assets.

Another example:
disk is 
```
d/e/file
```
Importing `d/` will get asset file without main asset `d/file` but with variant `d/e/file`
Importing `d/e/` is still unambiguous, asset `d/e/file` without a variant
But `d/**/` could mean either.

Pls advise.

**Comment by tvolkert** [comment]
I wrote up a sample of how this could work -- https://gist.github.com/tvolkert/40b0bcad883dd843f5829a881ddf82b1

Basically, it takes a raw listing of asset files (split out by package) and deduces which are the base assets versus the variants.

**Comment by fmatosqg** [comment]
@tvolkert thx for gist, I'll definitely be having a look.

First things first, it's hard for me to keep up the pace for lots of days and while globstar and subdir scanning is definitely a good thing, I got rejected on CI because of windows/posix, so I have to aim at there first. Hopefully me or somebody else can pick up from there.

The good news is that I got a way of making it work with `.posix` but `.windows` is breaking unit test everywhere. I guess having tests is good...

I had the same old "globals make test impossible" problem. I had to pass fs as parameter everywhere, and as far as I remember it was mostly limited to assets.dart. Is there any smarter way of replacing/overriding the `fs` global coming from `base/file__system.dart` with something being injected in the test, but still reverting to default value? 

Well, almost there.

**Comment by fmatosqg** [comment]
Amazing, extremely simplifies things :)
Last MemoryFileSystem question (for today lol),
How do I make everything run with both `.posix` and `.windows` on my laptop? So I know it won't fail on neither on both CIs.

I can think of some OO tricks with wrapping everything in a class and calling main twice. Or extending `testUsingContext` so it runs with 2 possible values for `fs`.

**Comment by tvolkert** [comment]
I'd create a `runTests(FileSystemStyle style)` method, like so:

```dart
void runTests(FileSystemStyle style) {
  group('...', () {
    MemoryFileSystem fs;

    setUp(() {
      fs = new MemoryFileSystem(style: style);
    });

    testUsingContext('...', () {
      ...
    }, overrides: <Type, Generator>{
      FileSystem: () => fs,
    });
  });
}
```

... then just call `runTests()` twice:

```dart
runTest(FileSystemStyle.posix);
runTest(FileSystemStyle.windows);
```

**Comment by fmatosqg** [comment]
I'm stuck at trying to load the schema file when switching to `MemoryFileSystem` with window style. I can't figure out a number of things, like

- what's the expected format for the string on `createSchemaFromUrl`, if starting with file:///C:/ or /C:/  (first one seems to work better when hardcoded
- how to generate such string. The beggining parts seem to be easy to find for POSIX but I can't figure a safe way to build it
- how to safely generate the file at the expected place `createSchemaFromUrl` will try to read when called from `flutter_manifest.dart` `_validate`. I can generate the "expected" folder and file either with `dart:io` or from the `MemoryFileSystem`, and I can successfully read from that file, although the contents are not what I expect. Though skipping either folder or file generation results in an error.

At best I get `FileSystemException: Cannot open file, path = '/C:/Users/fmatos/code/android/flutter/fwk_flutter/packages/flutter_tools/schema/pubspec_yaml.json' (OS Error: No such file or directory, errno = 2)` in many of those situations, coming from 
`return new File(filePath).readAsString()` on `createSchemaFromUrl`

If modifying `createSchemaFromUrl` to instead 
```
final hardcodedSchema = 'file:///Users/fmatos/code/android/flutter/fwk_flutter/packages/flutter_tools/schema/pubspec_yaml.json';
schemaUrl=hardcodedSchema;
```
makes everything work, but I'm afraid that io.File may not be using the right file system.

```
var editedSchemaPath = 'file://C:/Users/fmatos/code/android/flutter/fwk_flutter/packages/flutter_tools/schema/pubspec_yaml.json';
schemaUrl=editedSchemaPath;
``` 
also works.

**Comment by fmatosqg** [comment]
TL;DR for above post,

is it right to have the following unit test?
```
final uri = fs.path.toUri(schemaPath).toString();
assert(!uri.startsWith('file:///C:/'),'Uri for posix should be file://C:/');
```

where flutter_manifest has:
```
String get schemaPath => fs.path.join(
    fs.path.absolute(Cache.flutterRoot), 'packages', 'flutter_tools', 'schema',
    'pubspec_yaml.json',
  );

```


Currently that unit test will make sure we're compatible with `schema.dart`, but it's failing according to (I guess) something written in `MemoryFileSystem.dart`.

**Comment by fmatosqg** [comment]
Freezing work on MemFs for a while, making no progress for quite some time. Last almost stable version is on https://github.com/fmatosqg/flutter/tree/asset_scan_memfs  fbda4f14456a6b621cc28487988b66006ee5dbd1

Going back to original work on scanning paths according to agreed specs.

**Comment by fmatosqg** [comment]
Checklist:

- [x ]  implement non-recursive directory scan, write unit tests
- [ ]  get appveyor success
- [ ]  squash commits
- [ ]  code review
- [ ]  check for timestamp of pubspec, scanned files
- [ ]  implement globstar, write unit tests
- [ ]  write specs for fonts
- [ ]  write non-recursive directory scan for fonts, write unit tests

**Comment by tvolkert** [comment]
@fmatosqg that you for your hard work on this!  FYI, I'm probably not going to have time to review until at least Wednesday this week.

Just didn't want to give the impression I'm ignoring the review or ungrateful 😄

**Comment by fmatosqg** [comment]
Wed is probably before I'll be ready.

I'm starting to think it's unlikely to pass through windows tests without modifying other files such as `flutter_manifest.dart` or `package_map.dart`, they show very odd behaviour when used with `memory_filesystem.dart`.

I can write tests for memory_filesystem and others to expose my problems, but I'm not sure I'd know what's the expected behaviour. Things like who uses `file:///Users` or `/Users` are not clear to me, I'd appreciate if you can direct me some documentation. On windows there seems to be even more variations: `C:\` `/C:/` `file:///C:/` and weird combos of `/` and `\` in the same string and sometimes still working.

**Comment by fmatosqg** [comment]
Not ready for review. Needs extensive cleaning.

**Comment by fmatosqg** [comment]
@tvolkert I'd appreciate your feedback on the latest commit b3e83a1 for specifically 2 files that are part of the rabbit hole - `package_map.dart` and `flutter_manifest.dart`. My cleaning is close to finished in those 2 files, excepted obvious comments and `print`. But they have odd code that I have no idea how to make better. 

One thing that I can't figure out is how to detect it's a windows filesystem when on LocalFileSystem. I need that to feed `new Url.fromFile(...,windows:true)` in a couple of places.

Ideally I'd like to work on making those files close to "LGTM" before code reviewing `assets.dart` so we will have the chance to get a focused conversation when going there.

**Comment by fmatosqg** [comment]
Suggestion from gitter dart: use Platform when fs is LocalFileSystem. Then I'd use style when StyleableFileSystem.

**Comment by fmatosqg** [comment]
@tvolkert this is it, it's ready for code review. I'll push the squash with the first review change.

I'd be more comfortable wrapping this work without globstar for now, there's a couple of things I'd consider more critical such as adding new file to disk and re-generating bundle for hot reloads, or adding docs. Also because I'd like to take a break from this code.

Don't feel pressured, but I'd appreciate a signal in case you're busy but are aware of this state.

Thanks for the guidance on the critical steps, I totally understand that this is not a priority ticket and  everybody has other responsibilities.

**Comment by Hixie** [comment]
Hi, thanks for your contribution! @tvolkert will probably continue reviewing your PR next week but I was wondering if you could briefly describe what the rules you've implemented are since it wasn't quite clear to me from the patch what the intent was. Is the current rule just "if you specify a directory, we will include all the assets in that directory"?

**Comment by fmatosqg** [comment]
@Hixie thx for quick response.
That's correct, if an asset listed in `pubspec.yaml` ends in `/` it will be treated as a directory and all files included directly under it will be added as if listed directly. Variants are scanned accordingly. No support for recursive folders, globstar or wildcards.

I'll probably have docs edited in next few days along with an example app, I hope they help understand what it is about.

**Comment by fmatosqg** [comment]
I'm all for implementing whole subtree, but I think it's going to be complicated because of one edge case in variant support, as mentioned in one of the previous messages buried in this thread (https://github.com/flutter/flutter/pull/16413#issuecomment-380329565). In case it's needed before wrapping this PR I'd need clarification on what's expected for the examples mentioned there, so I know which unit tests to write.

**Comment by fmatosqg** [comment]
to clarify the ambiguous example with realistic file names

given the following files
```
./assets/1.png
./assets/2.0x/1.png
./assets/3.0x/1.png
./assets/3.0x/2.png
```

**Option A:**
I'd expect pubspec.yaml to be either
```
assets:
  - assets/1.png
  - assets/2.png
```
or
```
assets:
  - assets/
```

**Option B:**
However I'm afraid that naive whole subtree scanning will make

```
assets:
  - assets/
```
equivalent to
```
assets:
  - assets/1.png
  - assets/3.0x/2.png
```

which is clearly not the developer intention since it will take file 2.png which is high resolution targeting iPad and try to show in a size bigger than intended. 

If Option A is the intention I'm not sure what kind of heuristics to use, by dry-running the gist provided by @tvolkert I think it's going to implement Option B.

If Option B is desired I'd appreciate understanding it so we can get on the same page, but my opinion is that it's going to present UI problems when showing image at its natural size. Or if the gist implements Option A then I'd be glad to be wrong and move forward implementing it.

I notice I had a unit test for Option A that got lost in one of my refactors, I'll be getting it back later.

**Comment by fmatosqg** [comment]
Is Option B + globstar necessary to make this PR unanimously desirable? I'd be willing to move forward quickly if the answer is yes. 

I particularly think this would be skewing the sdk further away from native mobile practices but I'm ok if it helps to move on. The biggest problem for me is that I have little control of my free time to work on this and in a few weeks I won't have any - like most of the year.

**Comment by tvolkert** [comment]
Given the following files:

```
./assets/1.png
./assets/2.0x/1.png
./assets/3.0x/1.png
./assets/3.0x/2.png
```

And a pubspec of:

```
assets:
  - assets/1.png
  - assets/2.png
```

I'd expect there to be an error where we say we can't find `assets/2.png`.  Put another way, I'd expect that you need to have a base asset in order for the variant to be relevant.  As I say this though, I realize that I'm not sure if that's what happens today -- @fmatosqg do you know?

**Comment by Hixie** [comment]
@HansMuller may know what the current intent is and the history behind it, I think he was the last one to work on this.

**Comment by HansMuller** [comment]
As of https://github.com/flutter/flutter/pull/10901, it's not necessary to include the main asset, when resolution-specific variants are defined.  For example `./asssets/2.png' isn't strictly needed in https://github.com/flutter/flutter/pull/16413#issuecomment-386142790

**Comment by fmatosqg** [comment]
@tvolkert 
This is an example of what's supported today and how it may look like
https://github.com/fmatosqg/flutter_problem/tree/variants_PR_16413

Note that Option B.2 forces dev to reference the asset as `assets/3.0x/<filename>` in dart code.

image

**Comment by fmatosqg** [comment]
Also I'm assuming that mdpi files (1x) are less important given that they are rarely used - few phones or devices will pick them if a higher-res version is available. While xxhdpi (3x) are mandatory for all practical purposes: an app without it will suffer from low-res images in most phones.

**Comment by fmatosqg** [comment]
I also found a very curious case when fiddling with pubspec and checking why it didn't display as the size I expected: #17241. It's important to this ticket because `new Image.asset` disagrees with `AssetManifest.json`.

File: packages/flutter_tools/lib/src/asset.dart [hunk 1]
```diff
@@ -540,6 +552,71 @@ Map<_Asset, List<_Asset>> _parseAssets(
...
+  if (!fs.directory(directoryPath).existsSync()) {
+    printError('Error: unable to find directory entry in pubspec.yaml: $directoryPath');
+  }
+
+  final List<FileSystemEntity> lister = fs.directory(directoryPath)
+      .listSync(recursive: false);
+
+  for (FileSystemEntity entity in lister) {
+    if (entity is File) {
+      final String relPath = fs.path.relative(entity.path, from: assetBase);
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/asset.dart]: `relativePath`

File: packages/flutter_tools/lib/src/asset.dart [hunk 2]
```diff
@@ -540,6 +552,71 @@ Map<_Asset, List<_Asset>> _parseAssets(
...
+    { List<String> excludeDirs: const <String>[],
+      String packageName}) {
+  final String directoryPath = fs.path.join(
+      assetBase, assetUri.toFilePath(windows: platform.isWindows));
+
+  if (!fs.directory(directoryPath).existsSync()) {
+    printError('Error: unable to find directory entry in pubspec.yaml: $directoryPath');
+  }
+
+  final List<FileSystemEntity> lister = fs.directory(directoryPath)
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/asset.dart]: `recursive` defaults to `false`, so you can just inline this:

```dart
  for (FileSystemEntity entity in fs.directory(directoryPath).listSync()) {
```

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 3]
```diff
@@ -151,13 +152,27 @@ class FontAsset {
   String toString() => '$runtimeType(asset: ${assetUri.path}, weight; $weight, style: $style)';
 }
 
-Future<bool> _validate(Object manifest) async {
-  final String schemaPath = fs.path.join(
+@visibleForTesting
+String buildSchemaDir(FileSystem fs){
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: space after close parens

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 4]
```diff
@@ -151,13 +152,27 @@ class FontAsset {
...
+  return fs.path.join(
     fs.path.absolute(Cache.flutterRoot), 'packages', 'flutter_tools', 'schema',
+  );
+}
+
+@visibleForTesting
+String buildSchemaPath(FileSystem fs){
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: space after close parens

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 5]
```diff
@@ -151,13 +152,27 @@ class FontAsset {
...
+  return fs.path.join(
+    fs.path.absolute(buildSchemaDir(fs)),
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: `buildSchemaDir` already returns an absolute path, so this `fs.path.absolute()` shouldn't be needed

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 6]
```diff
@@ -151,13 +152,27 @@ class FontAsset {
...
     'pubspec_yaml.json',
   );
-  final Schema schema = await Schema.createSchemaFromUrl(fs.path.toUri(schemaPath).toString());
+}
+
+Future<bool> _validate(Object manifest) async {
+  final String schemaPath = buildSchemaPath(fs);
 
+  final String schemaData = fs.file(schemaPath).readAsStringSync();
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: Why was the change in how we load the schema necessary?

File: packages/flutter_tools/test/asset_bundle_package_test.dart [hunk 7]
```diff
@@ -48,13 +51,13 @@ $assetsSection
...
 
   void establishFlutterRoot() {
-    // Setting flutterRoot here so that it picks up the MemoryFileSystem's
-    // path separator.
     Cache.flutterRoot = getFlutterRoot();
   }
 
   void writePackagesFile(String packages) {
-    fs.file('.packages')
+    final Uri uri = new Uri.file('.packages', windows: platform.isWindows);
```
Comment (tvolkert) [inline packages/flutter_tools/test/asset_bundle_package_test.dart]: This shouldn't be necessary since file URIs only contain the `file:///` if they're absolute.  In this case, it'll just end up being `.packages`

File: packages/flutter_tools/test/asset_bundle_package_test.dart [hunk 8]
```diff
@@ -447,4 +457,244 @@ $assetsSection
...
+      final File schemaFile = filesystem.file(schemaPath);
+
+
+      final Directory schemaDir = filesystem.directory(
+          buildSchemaDir(filesystem));
+
+      schemaDir.createSync(recursive: true);
+      schemaFile.writeAsStringSync(schema);
+    }
+    //////////////////////////////////////
```
Comment (tvolkert) [inline packages/flutter_tools/test/asset_bundle_package_test.dart]: Just use a blank line here

File: packages/flutter_tools/test/asset_bundle_package_test.dart [hunk 9]
```diff
@@ -447,4 +457,244 @@ $assetsSection
...
+      const String _kFlutterRoot = '/flutter/flutter';
+      establishFlutterRoot();
+
+      final String schema = readSchemaPath(fs);
+
+      testUsingContext('$description - on windows FS', () async {
+        establishFlutterRoot();
+        writeSchema(schema, windowsFs);
+        await testMethod();
+      },
```
Comment (tvolkert) [inline packages/flutter_tools/test/asset_bundle_package_test.dart]:
```dart
}, overrides: <Type, Generator>{
  ...
});
```

File: packages/flutter_tools/test/asset_bundle_package_test.dart [hunk 10]
```diff
@@ -447,4 +457,244 @@ $assetsSection
...
+              'FLUTTER_ROOT': _kFlutterRoot,
+            },
+                operatingSystem: 'windows')
+          });
+
+      testUsingContext('$description - on posix FS', () async {
+        establishFlutterRoot();
+        writeSchema(schema, posixFs);
+        await testMethod();
+      },
```
Comment (tvolkert) [inline packages/flutter_tools/test/asset_bundle_package_test.dart]:
```dart
}, overrides: <Type, Generator>{
  ...
});
```

File: packages/flutter_tools/test/asset_bundle_package_test.dart [hunk 11]
```diff
@@ -447,4 +457,244 @@ $assetsSection
...
+      testUsingContext('$description - on windows FS', () async {
+        establishFlutterRoot();
+        writeSchema(schema, windowsFs);
+        await testMethod();
+      },
+          overrides: <Type, Generator>{
+            FileSystem: () => windowsFs,
+            Platform: () =>
+            new FakePlatform(environment: <String, String>{
+              'FLUTTER_ROOT': _kFlutterRoot,
```
Comment (tvolkert) [inline packages/flutter_tools/test/asset_bundle_package_test.dart]: Just put the environment map on one line to make the indentation easier to read.

File: packages/flutter_tools/test/asset_bundle_package_test.dart [hunk 12]
```diff
@@ -447,4 +457,244 @@ $assetsSection
...
+      testUsingContext('$description - on posix FS', () async {
+        establishFlutterRoot();
+        writeSchema(schema, posixFs);
+        await testMethod();
+      },
+          overrides: <Type, Generator>{
+            FileSystem: () => posixFs,
+            Platform: () =>
+            new FakePlatform(environment: <String, String>{
+              'FLUTTER_ROOT': _kFlutterRoot,
```
Comment (tvolkert) [inline packages/flutter_tools/test/asset_bundle_package_test.dart]: Just put the environment map on one line to make the indentation easier to read.

File: packages/flutter_tools/test/asset_bundle_package_test.dart [hunk 13]
```diff
@@ -447,4 +457,244 @@ $assetsSection
...
+        assetOnManifest,
+        <String>['test_package'],
+        expectedAssetManifest,
+      );
+    });
+
+    testUsingContextAndFs(
+        'Expect error generating manifest, wrong non-existing directory is listed', () async {
+      establishFlutterRoot();
+
```
Comment (tvolkert) [inline packages/flutter_tools/test/asset_bundle_package_test.dart]: Just one newline (same below)

**Review by tvolkert — COMMENTED** [review]
This looks great!  Almost all of my comments are just style nits.

File: packages/flutter_tools/lib/src/asset.dart [hunk 14]
```diff
@@ -496,27 +520,15 @@ Map<_Asset, List<_Asset>> _parseAssets(
...
-      final Uri relativeUri = fs.path.toUri(relativePath);
-      final Uri entryUri = asset.symbolicPrefixUri == null
-          ? relativeUri
-          : asset.symbolicPrefixUri.resolveUri(relativeUri);
-      variants.add(new _Asset(
-        baseDir: asset.baseDir,
-        entryUri: entryUri,
-        relativeUri: relativeUri,
-      ));
+    if ( assetUri.toString().endsWith('/')){
```
[the remark this answers is not in the document]
Comment (fmatosqg) [reply packages/flutter_tools/lib/src/asset.dart]: good catch, I used default dart formatter and it did as you say

File: packages/flutter_tools/lib/src/asset.dart [hunk 15]
```diff
@@ -485,6 +494,21 @@ class _AssetDirectoryCache {
 /// map of assets to asset variants.
 ///
 /// Returns null on missing assets.
+///
+/// Given package: 'test_package' and an assets directory like this:
```
[the remark this answers is not in the document]
Comment (fmatosqg) [reply packages/flutter_tools/lib/src/asset.dart]: Done 2 weeks ago: https://github.com/flutter/website/pull/992.
Not sure how to coordinate the merge, would you like me to remind you when this merge lands on beta so you can merge the `flutter/website`?

File: packages/flutter_tools/lib/src/asset.dart [hunk 16]
```diff
@@ -540,6 +552,71 @@ Map<_Asset, List<_Asset>> _parseAssets(
...
+        new _Asset(
+          baseDir: asset.baseDir,
+          entryUri: entryUri,
+          relativeUri: relativeUri,
+        )
+    );
+  }
+
+  result[asset] = variants;
+
```
[the remark this answers is not in the document]
Comment (fmatosqg) [reply packages/flutter_tools/lib/src/asset.dart]: not sure if extra was line before or after, I applied dart default formatter.

File: packages/flutter_tools/lib/src/asset.dart [hunk 17]
```diff
@@ -540,6 +552,71 @@ Map<_Asset, List<_Asset>> _parseAssets(
...
+  final List<_Asset> variants = <_Asset>[];
+  for (String path in cache.variantsFor(asset.assetFile.path)) {
+    final String relativePath = fs.path.relative(path, from: asset.baseDir);
+    final Uri relativeUri = fs.path.toUri(relativePath);
+    final Uri entryUri = asset.symbolicPrefixUri == null
+        ? relativeUri
+        : asset.symbolicPrefixUri.resolveUri(relativeUri);
+
+    variants.add(
+        new _Asset(
```
[the remark this answers is not in the document]
Comment (fmatosqg) [reply packages/flutter_tools/lib/src/asset.dart]: This one I had to edit manually, dart formatter would revert to commited version.

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 6 above)
Comment (fmatosqg) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: I couldn't find a way for the previous version to load the path from `MemoryFileSystem` on tests that used it. I believe it should work in production, but I wouldn't have a way to verify the windows behavior. 

If I remember correctly this was the line that caused problems `return new File(uri.scheme == 'file' ? uri.toFilePath() : schemaUrl)`, from `Schema.createSchemaFromUrl`. 

I did a lot of refactors from the 1st working version to the present cleaned one, but couldn't figure out a way to make `Schema.createSchemaFromUrl` work without changing it.

Comment (fmatosqg) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: BTW the tests I added on `flutter_manifest_test.dart` deal directly with this change.

**Comment by fmatosqg** [comment]
I did my best to follow all suggestions but by the number of them it was hard to make sure nothing slipped through.

Should I squash both commits and append comment `#4890` to reference issue number?

**Comment by tvolkert** [comment]
Thanks.  Don't worry about squashing commits - I'll squash when it's merged.

File: packages/flutter_tools/lib/src/asset.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (tvolkert) [reply packages/flutter_tools/lib/src/asset.dart]: Yeah, that'd be great.  thanks!

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 6 above)
Comment (tvolkert) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: Ah, I see now.  Yeah, it's pretty tricky -- you have to use `IOOverrides` when you want to inject your `MemoryFileSystem` into a package that doesn't use `package:file`.

Check out the following sample for how to do this in a test:

https://github.com/flutter/flutter/blob/f9bb4289e9fd861d70ae78bcc3a042ef1b35cc9d/packages/flutter_test/test/goldens_test.dart#L37-L57

With something like that, you should be able to keep the old call to `Schema.createSchemaFromUrl`, which should yield better asynchrony.

Comment (fmatosqg) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: There's a modification necessary on Schema (`schema.dart`) to make it work.

Consider the following method `futureSchema`, heavily based on `Schema.createSchemaFromUrl`: 

```
Future<Schema>  futureSchema(String schemaUrl) async {

  if (true || uri.scheme == 'file' || uri.scheme == '') {
    return new io.File(uri.scheme == 'file' ? uri.toFilePath() : schemaUrl)
        .readAsString()
        .then((text) => Schema.createSchema(convert.json.decode(text)));
  } else {
    throw new Exception('Nonsense c: is a valid uri scheme: $schemaUrl');
  }
}
```
Then we can keep the `_validate` method even simpler. 
```
Future<bool> _validate(Object manifest) async {
  final String schemaPath = buildSchemaPath(fs);
  final Schema schema = await futureSchema(schemaPath);
  final Validator validator = new Validator(schema);
  . . .
```

So far I couldn't find any combination of File,`new Uri`, `fs.path`, `fs.file` or `'file:// + ???` that could build a path that can be acceptable to the current version of `createSchemaFromUrl`. `MemoryFileSystem` is ok with `C:\....\....yaml` but the current checks (`if (uri.scheme == 'file' || uri.scheme == '')`) prevent it from working.

Comment (fmatosqg) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: here's another variation on `createSchemaFromUrl` that apparently would make it work: 
`final Uri uri = Uri.parse(r'$schemaUrl');`.

This whole effort is going beyond my reasonable energy.

**Comment by MaskyS** [comment]
@fmatosqg Thank you very much for your hard work on this. It is very much appreciated. For my use case, I have a lot of html files that I download from firebase based on user's choice, during app setup. Then I create a local jaguar server then serve these files. Without this feature, I'd have to create my own script that goes and finds all the files in a specific folder recursively then add those into pubspec.yaml. I hope this gets wrapped up in a couple of days!

**Comment by fmatosqg** [comment]
@MaskyS timeframes for this PR are better measured in weeks lol. That said I have figured out the last of @tvolkert's comments, but I'm quite short on time to clean it and roll it for all the places that need it.

Just a warning, the lastest version of the PR doesn't cover recursive scanning. That can be quite tricky because of some magic on how image with different resolutions are treated. 

That said, I find it very unorthodox to use more than a couple of htmls in the whole app, and even less ortohodox to have them bundled with the app during build.

**Comment by tvolkert** [comment]
@fmatosqg if you like, I can merge this PR and try to make the changes I'm suggesting in a follow-on change since they get quite deep in the weeds of Dart.  Would you like to proceed with that plan?

**Review by tvolkert — APPROVED** [review]
LGTM

**Comment by fmatosqg** [comment]
@tvolkert sounds awesome to me :) 

let me send you a commit that has everything figured out for `flutter_manifest_test.dart`, it's only pending applying the same "generate pubspec.yaml on the correct filesystem" for the other test file that covers `asset.dart`

**Comment by fmatosqg** [comment]
@tvolkert I appreciate the help. My latest code is here.  https://github.com/fmatosqg/flutter/tree/scan_assets_from_disk_needs_cleaning
- code will fail on analyzer
- `flutter_manifest_test.dart` should pass any unit tests
- `asset_bundle_package_test.dart` should fail only on tests that involve `MemoryFileSystem`, the fix I was thinking is make `testUsingContextAndFs` public (from `flutter_manifest_test.dart` or a new file) and simply reference them.

Reminder, pending checking/doing something on `asset.dart` `needsBuild()` for the situation where 
1.1 app is running with debug
1.2 developer adds/removes file to a scanned path (I'm almost sure it works but not sure on bad side effects)
2.1 full build
2.2 developer adds/removes file to scanned path
2.3. full build again, no clean. (observing `AssetManifest.json` and files copied to android intermediate folder will probably tell everything - I think it will work easier for adding files but removing them may be tricky)

**Comment by tvolkert** [comment]
Thanks again for the contribution and your patience with the long code review! 😄

**Comment by fmatosqg** [comment]
likewise 🥇

**Comment by mit-mit** [comment]
Thanks so much @fmatosqg for this very important improvement!
